File: OOP_Lab_1/Camera.cpp
#include "Camera.h"
#include <iostream>

Camera::Camera() : Equipment(), resolution("HD"), charge(0) {}

Camera::Camera(string n, string sn, string res, int ch) 
    : Equipment(n, sn), resolution(res), charge(ch) {}

Camera::Camera(const Camera& other) 
    : Equipment(other), resolution(other.resolution), charge(other.charge) {}

string Camera::getResolution() const { return resolution; }
void Camera::setResolution(string val) { resolution = val; }
int Camera::getCharge() const { return charge; }
void Camera::setCharge(int val) { charge = val; }

void Camera::takePhoto() {
    if (charge > 0) {
        cout << "  [*] Фото успешно сделано!" << endl;
        cout << "  [^] Заряд уменьшен на 1% (осталось: " << (charge-1) << "%)" << endl;
        charge--;
    } else {
        cout << "  [X] ОШИБКА: Батарея разряжена! Требуется подзарядка." << endl;
    }
}

void Camera::recordVideo() {
    cout << "  [#] Начата запись видео в разрешении " << resolution << endl;
}

void Camera::recharge() {
    int oldCharge = charge;
    charge = 100;
    cout << "  [+] Подзарядка завершена!" << endl;
    cout << "  [>] Было: " << oldCharge << "% -> Стало: " << charge << "%" << endl;
}

void Camera::work() {
    cout << endl;
    cout << "========================================" << endl;
    cout << "КАМЕРА: " << getName() << endl;
    cout << "========================================" << endl;
    cout << "  [>] Устройство начало работу..." << endl;
    takePhoto();
    cout << "  [OK] Камера готова к съемке" << endl;
}

string Camera::getDescription() {
    return "Тип: Камера | Название: " + getName() + 
           " | Серийный номер: " + getSerialNumber() + 
           " | Разрешение: " + resolution + 
           " | Заряд: " + to_string(charge) + "%";
}

File: OOP_Lab_3/Laba3OOP/src/main/java/com/equipment/entity/Camera.java
package com.equipment.entity;

import lombok.Data;
import javax.persistence.*;
import java.util.Arrays;
import java.util.List;

// Класс "Камера" - оборудование для съемки
@Entity
@Data
@Table(name = "cameras")
@PrimaryKeyJoinColumn(name = "equipment_id")
@DiscriminatorValue("Camera")
public class Camera extends Equipment {
    
    @Column
    private Integer resolution = 24; // разрешение в мегапикселях (значение по умолчанию)
    
    @Column
    private Integer charge = 100; // заряд батареи в процентах (значение по умолчанию)
    
    // КОНТРУКТОР ПО УМОЛЧАНИЮ 
    public Camera() {
        
    // КОНСТРУКТОР ПО ПАРАМЕТРАМ 
    }
    public Camera(String name, String serialNumber, StudioRoom studioRoom, 
                  Integer resolution, Integer charge) {
        super(name, serialNumber, studioRoom);
        this.resolution = resolution;
        this.charge = charge;
    }
    
    // КОНСТРУКТОР КОПИРОВАНИЯ 
    public Camera(Camera other) {
        super(other.getName() + " (Копия)", other.getSerialNumber() + "_copy_" + System.currentTimeMillis() + "_" + (int)(Math.random() * 1000), other.getStudioRoom());
        this.resolution = other.getResolution();
        this.charge = other.getCharge();
    }
    
    // Метод: сделать фото
    public String takePhoto() {
        if (charge > 0) {
            setCharge(charge - 1);
            return String.format("[*] Фото успешно сделано!\n[^] Заряд уменьшен на 1%% (осталось: %d%%)", charge);
        } else {
            return "[X] ОШИБКА: Батарея разряжена! Требуется подзарядка.";
        }
    }
    
    // Метод: начать запись видео
    public String recordVideo() {
        return String.format("[#] Начата запись видео в разрешении %d MP", resolution);
    }
    
    // Метод: зарядить батарею
    public String recharge() {
        int oldCharge = charge;
        setCharge(100);
        return String.format("[+] Подзарядка завершена!\n[>] Было: %d%% -> Стало: %d%%", oldCharge, charge);
    }
    
    @Override
    public String work() {
        String photoLog = takePhoto();
        return String.format("========================================\n" +
                   "КАМЕРА: %s\n" +
                   "========================================\n" +
                   "[>] Устройство начало работу...\n" +
                   "%s\n" +
                   "[OK] Камера готова к съемке", getName(), photoLog);
    }
    
    @Override
    public String getDescription() {
        return String.format("Тип: Камера | Название: %s | Серийный номер: %s | Разрешение: %d MP | Заряд: %d%%",
                getName(), getSerialNumber(), resolution, charge);
    }
    
    @Override
    public List<String> getAvailableActions() {
        return Arrays.asList("takePhoto", "recordVideo", "recharge");
    }
    
    @Override
    public String performAction(String action) {
        switch (action) {
            case "takePhoto":
                return takePhoto();
            case "recordVideo":
                return recordVideo();
            case "recharge":
                return recharge();
            case "work":
                return work();
            default:
                return "[X] Неизвестное действие: " + action;
        }
    }
    
    @Override
    public String toString() {
        return String.format("Camera{id=%d, name='%s', serialNumber='%s', " +
                        "resolution=%d MP, charge=%d%%, studioRoom='%s'}",
                getId(), getName(), getSerialNumber(),
                resolution, charge,
                getStudioRoom() != null ? getStudioRoom().getName() : "null");
    }
}
